### ISSUE TITLE
Explain ordering of subtables under arrays of tables

### ISSUE THREAD

**Issue by ChristianSi** [toml-lang/toml#658]
The spec says: "You can create nested arrays of tables as well. Just use the same double bracket syntax on sub-tables. Each double-bracketed sub-table will belong to the most recently defined table element above it."

This does not, however, explicitly mention that case that the subtable is a regular table rather than a table array -- though clearly the spec expects this case to be valid as well. This too should be mentioned and a suitable example added, e.g.

```toml
[[fruits]]
name = "pomegranate"

[fruits.tastes]
forbidden = true
sweet = true
spicy = false

[[fruits]]  # et cetera
name = "peach"
```

The spec also should make it clear that defining subtables (whether regular tables or arrays) *before* the first entry of a table array is not allowed, since it's not clear to which element of the (so far empty) table array they should be attached.

```
# INVALID
[fruits.tastes]  # this works only if "fruits" is a regular table
forbidden = true
sweet = true
spicy = false

[[fruits]]  # parser should throw an ERROR upon discovering that it's a table array
name = "pomegranate"

[[fruits]]  # et cetera
name = "peach"
```

Another issue about which the spec might be clarified are attempts to mix table arrays and plain tables. The spec already says: "Attempting to define a normal table with the same name as an already established array must produce an error at parse time", giving this example:

```toml
[[fruit.variety]]
  name = "red delicious"

# This table conflicts with the previous table
[fruit.variety]
  name = "granny smith"
```

While some might consider it obvious, it would be useful to add a "or vice versa" to make it clear that the reverse order is forbidden too:

```toml
[fruit.variety]
  name = "granny smith"

# This table conflicts with the previous table
[[fruit.variety]]
  name = "red delicious"
```

This issue is a spin-off of #651 – thanks @workingjubilee for bringing this to my attention.

@pradyunsg: Maybe add it to the critical path for TOML 1.0?

**Comment by mmakaay** [comment]
> The spec also should make it clear that defining subtables (whether regular tables or arrays) before the first entry of a table array is not allowed, since it's not clear to which element of the (so far empty) table array they should be attached.

Note that the actual thing that is happening there is something else.

```
# This is perfectly valid. It creates a table "fruits", which contains a table "tastes".
# There's no context that is about adding things to an array here. You simply create
# a standard table. Defining something in the spec that tells the reader that this is
# invalid is incorrect. This is valid, not invalid.
[fruits.tastes]
forbidden = true
sweet = true
spicy = false

# This is the place where things go wrong. This basically is the problem that you cannot
# redefine an existing table/table array. Creating [[fruits]] is invalid, because in the above
# code, "fruits" was already defined as a table and one cannot overwrite that table with 
# a table array.
[[fruits]]
name = "pomegranate"
```

To me, this seems the same kind of case as:

```
# VALID
fruits.tastes = "good"

[fruits] # INVALID

# or

[[fruits]] #INVALID
```

So the actual issue here is more generic and adding something specific to the spec to explain that you cannot create subtables before a table array feels wrong to me. It's not how you do it, that's for sure, but describing this situation as `[fruits.tastes]` being invalid might be really confusing.

**Comment by pradyunsg** [comment]
I'd rather say this is a case of -- a name used by an existing table/key can not be reused by an array.

**Comment by mmakaay** [comment]
Yes, that was exactly what I tried to convey, but failed to do it in just a few words like you, @pradyunsg  ;-) Thanks.

**Comment by ChristianSi** [comment]
@mmakaay: You're right that the error can only be detected the moment that `[[fruits]]` is defined -- something I also mentioned in my example.

Nevertheless, the two cases are not quite the same. If you define `fruits.tastes` as a dotted key, you define (at the same time) the "fruits" table, hence both `[fruits]` and `[[fruits]]` are not invalid (as you yourself pointed out). But if you define a `[fruits.tastes]` table, the parent table has not yet been defined, hence `[fruits]` remains valid.

So we cannot say that `[fruits.tastes]` implicitly defines the supertable -- it might be possible, however, to point out it creates an expectancy that "fruits" actually is a table (rather than an array or something else) provided that "fruits" wasn't explicitly defined earlier.

**Comment by AndrewSav** [comment]
Apologies, can some one remind me, why 

>If you define `fruits.tastes` as a dotted key, you define (at the same time) the "fruits" table,  but if you define a `[fruits.tastes]` table, the parent table has not yet been defined, hence `[fruits]` remains valid.

Just linking a comment where it has been addressed, why this makes sense is fine. I understand, that this is what the spec says, but why? Thank you in advance.

**Comment by ChristianSi** [comment]
@AndrewSav: I don't quite get your question, can you rephrase? What is it you don't understand?

**Comment by mmakaay** [comment]
Looking at recent developments in the spec, defining [fruits.tastes] defines both tables and defining [fruits] is no longer a possible valid table to define later on in the doc using [fruits].

**Comment by ChristianSi** [comment]
@mmakaay: You are mistaken. Ordering of table blocks doesn't matter and, of course, you may define any number of tables and subtables just as you like.

**Comment by AndrewSav** [comment]
>@AndrewSav: I don't quite get your question, can you rephrase? What is it you don't understand?

I'm a bit at a loss as to what's not clear.

You wrote (not verbatim):

>If you define fruits.tastes as a dotted key, you define (at the same time) the "fruits" table, but if you define a [fruits.tastes] table, the parent table has not yet been defined, hence [fruits] remains valid.

I'm asking, what is the rationale that defining a dotted key also define a parent table, but defining a "dotted table" table does NOT define said parent table. Hope this helps clear this up.

**Comment by mmakaay** [comment]
> @mmakaay: You are mistaken.

And you are not, I was terribly mistaken. I had seen the discussion about dotted keys in my mail and somehow misinterpreted the following from a recent commit:

"Dotted keys define everything to the left of each dot as a table. Since tables cannot be defined more than once, redefining such tables using a [table] header is not allowed. Likewise, using dotted keys to redefine tables already defined in [table] form is not allowed."

I read it as `[some.dotted.table.name]` being a dotted key as well, but that's a dotted table name instead. In a way, this represents the confusion that @AndrewSav  describes above. Intuitively, I saw a dotted key and a dotted table name as the same kind of thing, at the table-creating level.
